### PR TITLE
Foundation PR for adopting PHP-FPM-NGINX into ORYX

### PR DESCRIPTION
**Pull request by arroyc**

### CODE REVIEW

File: images/runtime/php-fpm/7.2/7.2.Dockerfile [hunk 1]
```diff
@@ -0,0 +1,246 @@
...
+# Adds GNU HASH segments to generated executables (this is used if present, and is much faster than sysv hash; in this configuration, sysv hash is also generated)
+# https://github.com/docker-library/php/issues/272
+# -D_LARGEFILE_SOURCE and -D_FILE_OFFSET_BITS=64 (https://www.php.net/manual/en/intro.filesystem.php)
+ENV PHP_CFLAGS="-fstack-protector-strong -fpic -fpie -O2 -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64"
+ENV PHP_CPPFLAGS="$PHP_CFLAGS"
+ENV PHP_LDFLAGS="-Wl,-O1 -Wl,--hash-style=both -pie"
+
+ENV GPG_KEYS 1729F83938DA44E27BA0F4D3DBDB397470D12172 B1B44D8F021E4E2D6021E995DC9FF8D3EE5AF27F
+
+ENV PHP_VERSION 7.2.25
```
Comment (kichalla) [inline images/runtime/php-fpm/7.2/7.2.Dockerfile]: I believe the version that is currently in master branch is 7.2.24. I would just stick to `24` for now so that this PR does not get huge.

File: images/runtime/php-fpm/7.3/7.3.Dockerfile [hunk 2]
```diff
@@ -0,0 +1,248 @@
...
+# Adds GNU HASH segments to generated executables (this is used if present, and is much faster than sysv hash; in this configuration, sysv hash is also generated)
+# https://github.com/docker-library/php/issues/272
+# -D_LARGEFILE_SOURCE and -D_FILE_OFFSET_BITS=64 (https://www.php.net/manual/en/intro.filesystem.php)
+ENV PHP_CFLAGS="-fstack-protector-strong -fpic -fpie -O2 -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64"
+ENV PHP_CPPFLAGS="$PHP_CFLAGS"
+ENV PHP_LDFLAGS="-Wl,-O1 -Wl,--hash-style=both -pie"
+
+ENV GPG_KEYS CBAF69F173A0FEA4B537F470D66C9593118BCCB6 F38252826ACD957EF380D39F2F7956BC5DA04B5D
+
+ENV PHP_VERSION 7.3.12
```
Comment (kichalla) [inline images/runtime/php-fpm/7.3/7.3.Dockerfile]: Also check this one with the one in master

File: images/runtime/php-fpm/buildRunTimeImageBases_Init.sh [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+
+baseImage="php-fpm-run-base"
+echo
+echo "Buildig image '$baseImage'..."
+docker build -t $baseImage -f "$CURRENT_DIR/runbase.Dockerfile" .
+
+for PHP_VERSION in "${VERSION_ARRAY[@]}"
+do
+	IFS='.' read -ra SPLIT_VERSION <<< "$PHP_VERSION"
+	VERSION_DIRECTORY="${SPLIT_VERSION[0]}.${SPLIT_VERSION[1]}"
```
Comment (kichalla) [inline images/runtime/php-fpm/buildRunTimeImageBases_Init.sh]: nit: fix indentation for all lines here

File: images/runtime/php-fpm/runbase.Dockerfile [hunk 4]
```diff
@@ -0,0 +1,55 @@
+FROM oryx-run-base
```
Comment (kichalla) [inline images/runtime/php-fpm/runbase.Dockerfile]: Is the content below this line from the official PHP github repository?

File: src/startupscriptgenerator/src/php/scriptgenerator.go [hunk 5]
```diff
@@ -29,8 +29,15 @@ func (gen *PhpStartupScriptGenerator) GenerateEntrypointScript() string {
 	scriptBuilder.WriteString("#!/bin/sh\n")
 	scriptBuilder.WriteString("# Enter the source directory to make sure the script runs where the user expects\n")
 	scriptBuilder.WriteString("cd " + gen.SourcePath + "\n")
-	scriptBuilder.WriteString("export APACHE_DOCUMENT_ROOT='" + gen.SourcePath + "'\n")
 	common.SetEnvironmentVariableInScript(&scriptBuilder, "APACHE_PORT", gen.BindPort, DefaultBindPort)
+	common.SetEnvironmentVariableInScript(&scriptBuilder, "NGINX_PORT", gen.BindPort, DefaultBindPort)
```
Comment (kichalla) [inline src/startupscriptgenerator/src/php/scriptgenerator.go]: Can you add end to end tests covering this scenario?

File: images/runtime/php-fpm/7.2/7.2.Dockerfile (hunk 1 above)
Comment (arroyc) [reply images/runtime/php-fpm/7.2/7.2.Dockerfile]: yeah in master we have 7.2.24, but php official library seems to have updated that to 7.2.25 on late November .. should we update as well or I should use 7.2.24 for fpm for now?

File: images/runtime/php-fpm/runbase.Dockerfile (hunk 4 above)
Comment (arroyc) [reply images/runtime/php-fpm/runbase.Dockerfile]: I have couple of extra libraries as following .. remaining are same
```
libzip-dev \
		libpng-dev \
		libjpeg-dev \
		libpq-dev \
		libldap2-dev \
		libldb-dev \
		libicu-dev \
		libgmp-dev \
		libmagickwand-dev \
		libc-client-dev \
		libtidy-dev \
		libkrb5-dev \
		libxslt-dev \
		unixodbc-dev \
		openssh-server \
		vim \
		curl \
		wget \
		tcptraceroute \
		mariadb-client \
		openssl \
```

File: images/runtime/php-fpm/7.3/7.3.Dockerfile (hunk 2 above)
Comment (arroyc) [reply images/runtime/php-fpm/7.3/7.3.Dockerfile]: master has 7.3.11 ..

File: images/runtime/php-fpm/7.2/7.2.Dockerfile (hunk 1 above)
Comment (kichalla) [reply images/runtime/php-fpm/7.2/7.2.Dockerfile]: Ok..let's change to `25` then

File: src/startupscriptgenerator/src/php/scriptgenerator.go (hunk 5 above)
Comment (arroyc) [reply src/startupscriptgenerator/src/php/scriptgenerator.go]: end to end tests are coming .. added some runtime

File: build/constants.yaml [hunk 6]
```diff
@@ -65,17 +66,18 @@
...
-    composer-version: 1.9.0
-    composer-sha384: a5c698ffe4b8e849a443b120cd5ba38043260d5c4023dbf93e1558871f1f07f58274fc6f4c93bcfd858c6bd0775cd8d1
+    php-build-base-tag: 20191214.3
+    php-runtime-base-tag: 20191214.3
+    php-fpm-runtime-base-tag: 20191214.2
+    composer-version: 1.9.1
+    composer-sha384: baf1608c33254d00611ac1705c1d9958c817a1a33bce370c0595974b342601bd80b92a3f46067da89e3b06bff421f182
     # hashes are for .tar.xz
-    php73-version: 7.3.11
+    php73-version: 7.3.12
```
Comment (dorfire) [inline build/constants.yaml]: can we do the version upgrades in a separate PR?
